File: python/024_Swap_Nodes_in_Pairs.py
"""
Given a linked list, swap every two adjacent nodes and return its head.

For example,
Given 1->2->3->4, you should return the list as 2->1->4->3.

Your algorithm should use only constant space. You may not modify the values in the list, only nodes itself can be changed.

Subscribe to see which companies asked this question
"""
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return head
        dummy = ListNode(-1)
        dummy.next = head
        tmp = dummy
        while tmp.next and tmp.next.next:
            thishead = tmp
            tmp = tmp.next
            thishead.next = tmp.next
            tmp.next = thishead.next.next
            thishead.next.next = tmp
        return dummy.next



class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

    @classmethod
    def create_from_array(cls, array):
        res = ListNode(-1)
        res_tmp = res
        for num in array:
            res_tmp.next = ListNode(num)
            res_tmp = res_tmp.next
        return res.next

    def display_list(self):
        tmp = self
        while tmp.next is not None:
            print tmp.val, 
            tmp = tmp.next
        print tmp.val


if __name__ == "__main__":
    head = ListNode.create_from_array([1,2,3,11,13])
    res = Solution()
    final = res.swapPairs(head)
    final.display_list()

File: java/024_Swap_Nodes_in_Pairs.java
// Given a linked list, swap every two adjacent nodes and return its head.

// For example,
// Given 1->2->3->4, you should return the list as 2->1->4->3.

// Your algorithm should use only constant space. You may not modify the values in the list, only nodes itself can be changed.


// Runtime: 5 ms


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode curNode = head;
        ListNode prevNode = dummy;
        while(curNode != null && curNode.next != null) {
            ListNode nextCurNode = curNode.next.next;
            curNode.next.next = curNode;
            prevNode.next = curNode.next;
            curNode.next = nextCurNode;
            prevNode = curNode;
            curNode = nextCurNode;
        }
        return dummy.next;
    }
}